chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```diff
@@ -81,11 +106,39 @@ case class ChainHandler(blockHeaderDAO: BlockHeaderDAO)(
     //this does _not_ mean that it is on the chain that has the most work
     //TODO: Enhance this in the future to return the "heaviest" header
     //https://bitcoin.org/en/glossary/block-chain
-    blockHeaderDAO.chainTips.map { tips =>
-      val sorted = tips.sortBy(header => header.blockHeader.difficulty)
-      val hash = sorted.head.hashBE
-      logger.debug(s"getBestBlockHash result: hash=$hash")
-      hash
+    val groupedChains = blockchains.groupBy(_.tip.height)
+    val maxHeight = groupedChains.keys.max
+    val chains = groupedChains(maxHeight)
+
+    val hashBE: DoubleSha256DigestBE = chains match {
+      case Vector() =>
+        val errMsg = s"Did not find blockchain with height $maxHeight"
+        logger.error(errMsg)
+        throw new RuntimeException(errMsg)
+      case chain +: Vector() =>
+        chain.tip.hashBE
+      case chain +: rest =>
+        logger.warn(
+          s"We have multiple competing blockchains: ${(chain +: rest).map(_.tip.hashBE.hex).mkString(", ")}")
+        chain.tip.hashBE
     }
+    Future.successful(hashBE)
+  }
+}
+
+object ChainHandler {
+
+  def fromDatabase(blockHeaderDAO: BlockHeaderDAO)(
+      implicit ec: ExecutionContext,
+      chainConfig: ChainAppConfig): Future[ChainHandler] = {
```

Comment:
Since this takes in these implicits, couldn't we just construct a `BlockHeaderDAO` from them?

Reply:
I would like to get away from passing around tons of implicits.